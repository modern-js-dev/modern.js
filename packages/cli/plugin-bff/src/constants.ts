export const API_APP_NAME = '_app';

export const BUILD_FILES = [
  '**/*.[tj]sx?',
  '!**/*.test.jsx?',
  '!**/*.test.tsx?',
  '!**/*.spec.jsx?',
  '!**/*.spec.tsx?',
  '!__tests__/*.tsx?',
  '!__tests__/*.jsx?',
];
