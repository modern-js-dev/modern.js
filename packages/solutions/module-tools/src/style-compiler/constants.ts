import { Option } from './types';

export const CSS_EXTENSIONS = ['.css', '.less', '.sass', '.scss'];

export const CSS_EXTENSIONS_RULE = {
  cssRule: /\.css$/,
  cssModuleRule: /\.module\.css$/,
  lessRule: /\.less$/,
  lessModuleRule: /\.module\.less$/,
  sassRule: /\.(scss|sass)$/,
  sassModuleRule: /\.module\.(scss|sass)$/,
};

export const defaultCompilerOptions: Option = {
  less: {},
  sass: {
    file: '',
  },
  postcss: { options: {} },
};
