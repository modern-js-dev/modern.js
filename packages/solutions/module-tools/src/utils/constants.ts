// generator tsconfig file
export const tempTsconfigName = 'tsconfig.temp.json';
