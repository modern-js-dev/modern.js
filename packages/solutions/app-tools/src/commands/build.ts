import { Configuration, webpack } from 'webpack';
import { WebpackConfigTarget, getWebpackConfig } from '@modern-js/webpack';
import { manager, PluginAPI, ResolvedConfigContext } from '@modern-js/core';
import {
  formatWebpackMessages,
  measureFileSizesBeforeBuild,
  printFileSizesAfterBuild,
  printBuildError,
  logger,
  isUseSSRBundle,
  emptyDir,
} from '@modern-js/utils';
import { generateRoutes } from '../utils/routes';
import { buildServerConfig, emitResolvedConfig } from '../utils/config';
import type { BuildOptions } from '../utils/types';

// These sizes are pretty large. We'll warn for bundles exceeding them.
const WARN_AFTER_BUNDLE_GZIP_SIZE = 512 * 1024;
const WARN_AFTER_CHUNK_GZIP_SIZE = 1024 * 1024;

export const build = async (api: PluginAPI, options?: BuildOptions) => {
  const resolvedConfig = api.useResolvedConfigContext();
  const appContext = api.useAppContext();
  const hookRunners = api.useHookRunners();
  const { existSrc } = appContext;

  if (!existSrc) {
    const { distDirectory } = appContext;
    await emptyDir(distDirectory);
    await hookRunners.beforeBuild({
      webpackConfigs: [],
    });

    await generateRoutes(appContext);

    await hookRunners.afterBuild();

    return;
  }

  const webpackBuild = async (webpackConfig: Configuration, type?: string) => {
    const compiler = webpack(webpackConfig);

    return new Promise((resolve, reject) => {
      let label = process.env.NODE_ENV || '';
      if (type && type !== 'legacy') {
        label += ` ${type}`;
      }
      logger.info(`Creating a ${label} build...`);

      compiler.run((err, stats) => {
        let messages: {
          errors: any;
          warnings: any;
        };
        if (!err) {
          messages = formatWebpackMessages(
            stats!.toJson({ all: false, warnings: true, errors: true }),
          );

          if (messages.errors.length === 0) {
            logger.info(`File sizes after ${label} build:\n`);
            printFileSizesAfterBuild(
              stats,
              previousFileSizes,
              distDirectory,
              WARN_AFTER_BUNDLE_GZIP_SIZE,
              WARN_AFTER_CHUNK_GZIP_SIZE,
            );
            logger.log();
          }
        }

        // When using run or watch, call close and wait for it to finish before calling run or watch again.
        // Concurrent compilations will corrupt the output files.
        compiler.close(closeErr => {
          if (closeErr) {
            logger.error(closeErr);
          }
          if (err) {
            reject(err);
          } else {
            if (messages.errors.length) {
              reject(new Error(messages.errors.join('\n\n')));
              return;
            }
            resolve({ warnings: messages.warnings });
          }
        });
      });
    });
  };

  manager.run(() => {
    ResolvedConfigContext.set({ ...resolvedConfig, cliOptions: options });
  });

  const { distDirectory, appDirectory, serverConfigFile } = appContext;
  const previousFileSizes = await measureFileSizesBeforeBuild(distDirectory);
  await emptyDir(distDirectory);

  await buildServerConfig(appDirectory, serverConfigFile);

  const buildConfigs: Array<{ type: string; config: any }> = [];
  buildConfigs.push({
    type: 'legacy',
    config: getWebpackConfig(WebpackConfigTarget.CLIENT)!,
  });

  if (resolvedConfig.output.enableModernMode) {
    buildConfigs.push({
      type: 'modern',
      config: getWebpackConfig(WebpackConfigTarget.MODERN)!,
    });
  }

  if (isUseSSRBundle(resolvedConfig)) {
    buildConfigs.push({
      type: 'ssr',
      config: getWebpackConfig(WebpackConfigTarget.NODE)!,
    });
  }

  await hookRunners.beforeBuild({
    webpackConfigs: buildConfigs.map(({ config }) => config),
  });

  for (const buildConfig of buildConfigs) {
    const { type: buildType, config } = buildConfig;
    try {
      await webpackBuild(config, buildType);
    } catch (error) {
      printBuildError(error as Error);
      // eslint-disable-next-line no-process-exit
      process.exit(1);
    }
  }

  await generateRoutes(appContext);
  await hookRunners.afterBuild();
  await emitResolvedConfig(appDirectory, resolvedConfig);
};
