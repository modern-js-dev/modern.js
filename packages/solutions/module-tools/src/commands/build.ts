import * as path from 'path';
import { fs, Import, dotenv } from '@modern-js/utils';
import type { PluginAPI } from '@modern-js/core';
import onExit from 'signal-exit';
import { tempTsconfigName } from '../utils/constants';
import type { Platform } from '../types';

const tsConfigutils: typeof import('../utils/tsconfig') = Import.lazy(
  '../utils/tsconfig',
  require,
);

const valid: typeof import('../utils/valide') = Import.lazy(
  '../utils/valide',
  require,
);
const buildFeature: typeof import('../features/build') = Import.lazy(
  '../features/build',
  require,
);
/**
 * init work before build task.
 * @param api
 */
export const init = (api: PluginAPI): void => {
  const { appDirectory } = api.useAppContext();

  dotenv.config();

  onExit(() => {
    const tempTsconfigFileAbsPath = path.join(
      appDirectory,
      `./${tempTsconfigName}`,
    );
    fs.removeSync(tempTsconfigFileAbsPath);
  });
};

export interface IBuildCommandOption {
  watch: boolean;
  tsconfig: string;
  platform: boolean | Exclude<Platform, 'all'>;
  styleOnly: boolean;
  tsc: boolean;
  clear: boolean;
}

export const build = async (
  api: PluginAPI,
  buildCommandOption: IBuildCommandOption,
) => {
  const {
    watch = false,
    tsconfig: tsconfigName,
    tsc,
    clear = true,
    platform,
  } = buildCommandOption;

  init(api);

  const { appDirectory } = api.useAppContext();
  const modernConfig = api.useResolvedConfigContext();
  const tsconfigPath = path.join(appDirectory, tsconfigName);
  const outputPath = modernConfig.output.path ?? 'dist';
  const isTsProject = tsConfigutils.existTsConfigFile(tsconfigPath);
  const enableTscCompiler =
    isTsProject && tsc && !modernConfig.output.disableTsChecker;

  valid.valideBeforeTask({ modernConfig, tsconfigPath });

  // TODO: 一些配置只需要从modernConfig中获取
  await buildFeature.build(
    api,
    {
      appDirectory,
      enableWatchMode: watch,
      isTsProject,
      platform,
      sourceDir: 'src',
      tsconfigName,
      enableTscCompiler,
      clear,
      outputPath,
    },
    modernConfig,
  );
};
