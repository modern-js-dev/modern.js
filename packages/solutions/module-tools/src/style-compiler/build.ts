import { CSS_EXTENSIONS, defaultCompilerOptions } from './constants';
import { resolveFiles } from './resolveFiles';
import { getFiles } from './utils';
import { IBuildOption } from './types';

export { CSS_EXTENSIONS };

export async function transformStyle(option: IBuildOption) {
  const {
    projectDir,
    stylesDir,
    outDir,
    enableVirtualDist = false,
    compilerOption = defaultCompilerOptions,
  } = option;
  // get style files
  const files = getFiles({
    stylesDir,
    extensions: CSS_EXTENSIONS,
  });

  if (enableVirtualDist) {
    return resolveFiles({
      files,
      projectDir,
      stylesDir,
      outDir,
      enableVirtualDist,
      options: compilerOption,
    });
  } else {
    await resolveFiles({
      files,
      projectDir,
      stylesDir,
      outDir,
      enableVirtualDist,
      options: compilerOption,
    });
    return undefined;
  }
}
