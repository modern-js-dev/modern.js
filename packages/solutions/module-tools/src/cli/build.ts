import { Command } from '@modern-js/utils';
import type { PluginAPI } from '@modern-js/core';
import type { IBuildCommandOption } from '../commands/build';
import { i18n, localeKeys } from '../locale';
import { build } from '../commands/build';

export const buildCli = (program: Command, api: PluginAPI) => {
  return (
    program
      .command('build')
      .usage('[options]')
      .description(i18n.t(localeKeys.command.build.describe))
      .option('-w, --watch', i18n.t(localeKeys.command.build.watch))
      .option(
        '--tsconfig [tsconfig]',
        i18n.t(localeKeys.command.build.tsconfig),
        './tsconfig.json',
      )
      .option('--style-only', i18n.t(localeKeys.command.build.style_only))
      .option(
        '-p, --platform [platform]',
        i18n.t(localeKeys.command.build.platform),
      )
      // @deprecated
      // The `--no-tsc` option has been superceded by the `--no-dts` option.
      .option('--no-tsc', i18n.t(localeKeys.command.build.no_tsc))
      .option('--dts', i18n.t(localeKeys.command.build.dts))
      .option('--no-clear', i18n.t(localeKeys.command.build.no_clear))
      .option('-c --config <config>', i18n.t(localeKeys.command.build.config))
      .action(async (subCommand: IBuildCommandOption) => {
        await build(api, subCommand);
      })
  );
};
