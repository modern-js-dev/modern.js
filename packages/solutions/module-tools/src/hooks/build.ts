import { createParallelWorkflow, createAsyncPipeline } from '@modern-js/plugin';
import {
  registerHook,
  NormalizedConfig,
  LessOption,
  SassOption,
  PostcssOption,
} from '@modern-js/core';
import { CompilerItem, PostcssCompilerItem } from '../style-compiler';

export interface PlatformBuildOption {
  isTsProject: boolean;
}

export const platformBuild = createParallelWorkflow<
  PlatformBuildOption,
  {
    name: string;
    title: string;
    taskPath: string;
    params: string[];
  }
>();

export const moduleLessConfig = createAsyncPipeline<
  { modernConfig: NormalizedConfig },
  LessOption | undefined
>();

export const moduleSassConfig = createAsyncPipeline<
  { modernConfig: NormalizedConfig },
  SassOption | undefined
>();

export const modulePostcssConfig = createAsyncPipeline<
  { modernConfig: NormalizedConfig; appDirectory: string },
  PostcssOption | undefined
>();

export const moduleTailwindConfig = createAsyncPipeline<
  { modernConfig: NormalizedConfig },
  any
>();

export const getModuleLessCompiler = createAsyncPipeline<
  unknown,
  CompilerItem | undefined
>();

export const getModuleSassCompiler = createAsyncPipeline<
  unknown,
  CompilerItem | undefined
>();

export const getModulePostcssCompiler = createAsyncPipeline<
  unknown,
  PostcssCompilerItem | undefined
>();

export const buildHooks = {
  platformBuild,
  moduleLessConfig,
  getModuleLessCompiler,
  moduleSassConfig,
  getModuleSassCompiler,
  modulePostcssConfig,
  getModulePostcssCompiler,
  moduleTailwindConfig,
};

export const lifecycle = () => {
  registerHook({
    moduleLessConfig,
    getModuleLessCompiler,
    moduleSassConfig,
    getModuleSassCompiler,
    modulePostcssConfig,
    getModulePostcssCompiler,
    moduleTailwindConfig,
    platformBuild,
  });
};

declare module '@modern-js/core' {
  export interface Hooks {
    platformBuild: typeof platformBuild;
    moduleLessConfig: typeof moduleLessConfig;
    getModuleLessCompiler: typeof getModuleLessCompiler;
    moduleSassConfig: typeof moduleSassConfig;
    getModuleSassCompiler: typeof getModuleSassCompiler;
    modulePostcssConfig: typeof modulePostcssConfig;
    getModulePostcssCompiler: typeof getModulePostcssCompiler;
    moduleTailwindConfig: typeof moduleTailwindConfig;
  }
}
