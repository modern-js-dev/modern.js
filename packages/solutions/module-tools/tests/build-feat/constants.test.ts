import { unPresetWithTargetConfigs } from '../../src/features/build/constants';

describe('constants', () => {
  test('unPresetWithTargetConfigs', () => {
    expect(unPresetWithTargetConfigs).toMatchSnapshot();
  });
});
