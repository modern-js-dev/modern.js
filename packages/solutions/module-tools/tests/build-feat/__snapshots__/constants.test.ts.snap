// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constants unPresetWithTargetConfigs 1`] = `
Object {
  "npm-component-es2015": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2015",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2015",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2015",
    },
  ],
  "npm-component-es2016": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2016",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2016",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2016",
    },
  ],
  "npm-component-es2017": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2017",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2017",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2017",
    },
  ],
  "npm-component-es2018": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2018",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2018",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2018",
    },
  ],
  "npm-component-es2019": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2019",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2019",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2019",
    },
  ],
  "npm-component-es2020": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2020",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2020",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2020",
    },
  ],
  "npm-component-es5": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es5",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es5",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es5",
    },
  ],
  "npm-component-es6": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es6",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es6",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es6",
    },
  ],
  "npm-component-esnext": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "esnext",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "esnext",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "esnext",
    },
  ],
  "npm-component-with-umd-es2015": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2015",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2015",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2015",
    },
  ],
  "npm-component-with-umd-es2016": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2016",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2016",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2016",
    },
  ],
  "npm-component-with-umd-es2017": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2017",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2017",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2017",
    },
  ],
  "npm-component-with-umd-es2018": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2018",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2018",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2018",
    },
  ],
  "npm-component-with-umd-es2019": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2019",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2019",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2019",
    },
  ],
  "npm-component-with-umd-es2020": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2020",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2020",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2020",
    },
  ],
  "npm-component-with-umd-es5": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es5",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es5",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es5",
    },
  ],
  "npm-component-with-umd-es6": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es6",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es6",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es6",
    },
  ],
  "npm-component-with-umd-esnext": Array [
    Object {
      "buildType": "bundleless",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "esnext",
    },
    Object {
      "buildType": "bundleless",
      "format": "esm",
      "outputPath": "./es",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "esnext",
    },
    Object {
      "buildType": "bundleless",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "esnext",
    },
  ],
  "npm-library-es2015": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2015",
    },
  ],
  "npm-library-es2016": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2016",
    },
  ],
  "npm-library-es2017": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2017",
    },
  ],
  "npm-library-es2018": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2018",
    },
  ],
  "npm-library-es2019": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2019",
    },
  ],
  "npm-library-es2020": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2020",
    },
  ],
  "npm-library-es5": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es5",
    },
  ],
  "npm-library-es6": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es6",
    },
  ],
  "npm-library-esnext": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "esnext",
    },
  ],
  "npm-library-with-umd-es2015": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2015",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2015",
    },
  ],
  "npm-library-with-umd-es2016": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2016",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2016",
    },
  ],
  "npm-library-with-umd-es2017": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2017",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2017",
    },
  ],
  "npm-library-with-umd-es2018": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2018",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2018",
    },
  ],
  "npm-library-with-umd-es2019": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2019",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2019",
    },
  ],
  "npm-library-with-umd-es2020": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es2020",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es2020",
    },
  ],
  "npm-library-with-umd-es5": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es5",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es5",
    },
  ],
  "npm-library-with-umd-es6": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "es6",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "es6",
    },
  ],
  "npm-library-with-umd-esnext": Array [
    Object {
      "buildType": "bundle",
      "format": "cjs",
      "outputPath": "./lib",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "format": "esm",
      "outputPath": "./es",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "format": "umd",
      "outputPath": "./umd",
      "target": "esnext",
    },
    Object {
      "buildType": "bundle",
      "dtsOnly": true,
      "enableDts": true,
      "outputPath": "./types",
      "target": "esnext",
    },
  ],
}
`;
